### ISSUE TITLE
Thoughts on fixing semantic conflict issues for common functions

### ISSUE THREAD

**Issue by PHILO-HE** [facebookincubator/velox#4876]
### Description

As we know, for presto function under `prestosql` folder, if there are some semantic conflict issues for spark, we generally fix them by re-implementing another function under `sparksql` folder.

Recently, we found there are some semantic conflict issues in some common functions, like `cast`, e.g., spark allows decimal string, like `123.5`, to be cast to int type, but currently velox produces null for such case. This issue can be fixed by introducing a config to let user to control the expected behavior. I note velox has a config called `kCastIntByTruncate` to fix a similar issue. But if multiple configs are introduced, it makes the code complex and hard to maintain. Another thought is, we can separate the implementations, like what we did under `sparksql` & `prestosql` folders. Definitely, such approach will cause a lot of repeated code.

@mbasmanova @pedroerp, do you have any idea? Thanks!

**Comment by mbasmanova** [comment]
@PHILO-HE I'm seeing that Presto supports casting decimal to integer:

```
presto:di> select cast(decimal '123.4' as integer);
 _col0
-------
   123
(1 row)
```

Hence, we should just extend cast to add such support. CC: @majetideepak @karteekmurthys 

Overall, I'd suggest to continue introducing configuration properties to alter behavior of special forms like CAST as needed.

**Comment by PHILO-HE** [comment]
@mbasmanova, thanks for your comment!
Let me clarify a bit. The mentioned example is to cast a string which represents a decimal number, not about decimal type.

If we continue using configuration properties, do we need to cover every if/else branch? See the below possible example, where P1 & P2 are two config properties to be introduced. Or, just cover the if/else branch we really need in spark or presto.
```
if P1 is true {
   ...
   if P2 is true {
     ...
  } else {
     ...
  }
} else {
  ...
  if P2 is true {
   ..
  } else {
   ..
  }
}
```

**Comment by mbasmanova** [comment]
> The mentioned example is to cast a string which represents a decimal number, not about decimal type.

Thank you for clarification. Presto supports casting such string to decimal. Does Spark support casting to integer?

```
presto:di> select cast('123.4' as decimal);
 _col0
-------
 123
(1 row)

presto:di> select cast('123.4' as integer);
Query 20230509_151255_34922_tkqte failed: Cannot cast '123.4' to INT
```

> If we continue using configuration properties, do we need to cover every if/else branch?

I think the answer would depend on specific properties. We'll need to look into this on a case by case basis.

**Comment by PHILO-HE** [comment]
> Thank you for clarification. Presto supports casting such string to decimal. Does Spark support casting to integer?

@mbasmanova, yes, spark supports casting it to integer, the result is `123` for your example.

I think we can propose some fixes still in the config approach and have more discussion on specific cases. Thanks!

**Comment by PHILO-HE** [comment]
Hi @mbasmanova, it seems there is no spark doc to describe the semantics for cast functions. I will try to summarize one. Thanks!

**Comment by mbasmanova** [comment]
@PHILO-HE Thanks. Having a doc describing supports CASTs and their semantics would allow us to decide whether to create a Spark-specific CAST implementation or adjust existing implementation with more flags.

**Comment by JkSelf** [comment]
@mbasmanova @PHILO-HE 
when performing the cast(2.56 Decimal(6, 2) as int) in Velox, the result would be 3, while Spark would return 2. Filed https://github.com/facebookincubator/velox/pull/6208 to support truncate the fraction part of a decimal number when casting decimal to an integer.

**Comment by pedroerp** [comment]
@mbasmanova @PHILO-HE given the many corner case differences between the cast behavior between these engines, I wonder if we should provide a nicer way for engines to specialize the cast code instead of relying on flags. Like scalar functions, maybe we could provide a baseline cast implementation (the base CastExpr class) following, say, Presto semantic, and let engines overwrite its specific conversion methods?

**Comment by mbasmanova** [comment]
@pedroerp This might be a good idea. Maybe someone can prototype this to see what it looks like.

**Comment by PHILO-HE** [comment]
There are a lot of cases listed in the below link for demonstrating presto's cast behavior. Will do a check for spark.
https://github.com/facebookincubator/velox/blob/main/velox/docs/functions/presto/conversion.rst

**Comment by rui-mo** [comment]
@mbasmanova Opened https://github.com/facebookincubator/velox/pull/7377 for the base layout of Spark cast expr. Could you take a look? Thanks.

**Comment by rui-mo** [comment]
Below is a table summarizing the differences between Spark and Presto CAST implementations we noticed. Some of them have been solved via config, like `cast_to_int_by_truncate` and `cast_string_to_date_is_iso_8601`.

| Conversion |  Cases | Presto result | Spark result
| ------------- | ------------- |  ------------- | ------------ |
| floating-point ->  integral | SELECT cast(12345.67 as bigint); | 12346 | 12345 |
|  | SELECT cast(127.8 as tinyint);  | out of range |  127 |
|  | SELECT cast(129.9 as tinyint);  | out of range | -127 |
|  | SELECT cast(infinity() as bigint); | out of range | 9223372036854775807 |
| decimal -> integral | SELECT cast(2.56 decimal(6, 2) as integer); | 3 | 2 |
|  | SELECT cast(decimal '300.001' as tinyint);  | out of range | 44 |
| string -> integral | SELECT cast('12345.67' as int); | invalid | 12345 |
| string -> decimal | select cast('&nbsp;&nbsp;1.23&nbsp;&nbsp;' as decimal(38, 0)); | invalid | 123 |
| string -> date | select cast('2015' as date); | invalid | 2015-01-01 |
|  | and some other string patterns supported by Spark |  |  |
| timestamp -> string | select cast(cast('2015-03-18 12:03:17' as timestamp) as varchar);  | 2015-03-18 12:03:17.000 | 2015-03-18 12:03:17 |
| string -> timestamp | select cast('2015-03-18T12:03:17' as timestamp) | invalid | 2015-03-18 12:03:17 |
|  | select cast('0001' as timestamp) | invalid | 0001-01-01 00:00:00 |
|  | and some other string patterns supported by Spark |  |  |
